File: algorithms/algorithms-python/leetcode/Question_0079_Word_Search.py
#!/usr/bin/python
# coding: utf-8

class Solution(object):
    dr = [1, 0, -1, 0]
    dc = [0, 1, 0, -1]
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        m, n = len(board), len(board[0])
        visited = [[False] * n for i in range(m)]
        for i in range(m):
            for j in range(n):
                if self.isThisWay(board, word, i, j, 0, visited):
                    return True
        return False

    def isThisWay(self, board, word, row, col, index, visited):
        if row < 0 or row >= len(board) or col < 0 or col >= len(board[0]) or \
            visited[row][col] or board[row][col] != word[index]:
            return False
        
        index += 1
        if index == len(word):
            return True
        visited[row][col] = True
        for i in range(4):
            if self.isThisWay(board, word, row + self.dr[i], col + self.dc[i], index, visited):
                return True
        visited[row][col] = False
        return False

File: algorithms/algorithms-cplusplus/leetcode/Question_0079_Word_Search.cc
#include <vector>
#include <string>

using namespace std;

class Solution {
 public:
  bool exist(vector<vector<char>>& board, string word) {
    vector<vector<bool>> visited(board.size(), vector<bool>(board[0].size()));
    for (int i = 0; i < board.size(); i++) {
      for (int j = 0; j < board[0].size(); j++) {
        if (isThisWay(board, word, i, j, 0, visited)) {
          return true;
        }
      }
    }
    return false;
  }

  bool isThisWay(vector<vector<char>>& board, string& word, int row,
      int column, int index, vector<vector<bool>>& visited) {
    if (row < 0 || row >= board.size() || column < 0 || column >= board[0].size()
        || visited[row][column] || board[row][column] != word[index]) {
      return false;
    }
  
    if (++index == word.size()) {
      return true;
    }
    visited[row][column] = true;
    if (isThisWay(board, word, row - 1, column, index, visited) ||
        isThisWay(board, word, row + 1, column, index, visited) ||
        isThisWay(board, word, row, column - 1, index, visited) ||
        isThisWay(board, word, row, column + 1, index, visited)) {
      return true;
    }
    visited[row][column] = false;
    return false;
  }
};
